src/Query/QueryBuilder.php

```diff
@@ -557,6 +601,33 @@ public function addUnion(string|QueryBuilder $part, UnionType $type = UnionType:
         return $this;
     }
 
+    /**
+     * Add a Common Table Expression to be used for a select query.
+     *
+     * <code>
+     *     // WITH cte_name AS (SELECT 1 AS column1)
+     *     $qb = $conn->createQueryBuilder()
+     *         ->with('cte_name', 'SELECT 1 AS column1');
+     *
+     *     // WITH cte_name(column1) AS (SELECT 1 AS column1)
+     *     $qb = $conn->createQueryBuilder()
+     *         ->with('cte_name', 'SELECT 1 AS column1', ['column1']);
+     * </code>
+     *
+     * @param string   $name    The name of the CTE
+     * @param string[] $columns The optional columns list to select in the CTE.
+     *
+     * @return $this This QueryBuilder instance.
+     */
+    public function with(string $name, string|QueryBuilder $part, array $columns = []): self
```

Comment:
Should we make the default `$colums` value `null` and make the empty array invalid? Otherwise, if someone computes the value of `$columns` programmatically and ends up with an empty array, it will result in selecting all columns. In general, the semantics "empty means everything/no-limit" is a footgun.